esp-hal/src/uart.rs

```diff
@@ -904,24 +902,80 @@ where
             .write(|w| unsafe { w.clkdiv().bits(divider).frag().bits(0) });
     }
 
-    #[cfg(any(esp32c6, esp32h2))] // TODO introduce a cfg symbol for this
+    #[cfg(esp32c3)]
     #[inline(always)]
-    fn sync_regs(&self) {
+    fn init() {
+        let system = unsafe { crate::peripherals::SYSTEM::steal() };
+        if !system.perip_clk_en0().read().uart_mem_clk_en().bit() {
+            system
+                .perip_clk_en0()
+                .modify(|_, w| w.uart_mem_clk_en().set_bit());
+        }
+
+        system
+            .perip_clk_en0()
+            .modify(|_, w| match T::uart_number() {
+                0 => w.uart_clk_en().set_bit(),
+                1 => w.uart1_clk_en().set_bit(),
+                _ => panic!("Unknown Uart Device!"),
+            });
+        system
+            .perip_rst_en0()
+            .modify(|_, w| match T::uart_number() {
+                0 => w.uart_rst().clear_bit(),
+                1 => w.uart1_rst().clear_bit(),
+                _ => panic!("Unknown Uart Device!"),
+            });
+
+        T::register_block()
+            .clk_conf()
+            .modify(|_, w| w.rst_core().set_bit());
+        system
+            .perip_rst_en0()
+            .modify(|_, w| match T::uart_number() {
+                0 => w.uart_rst().set_bit(),
+                1 => w.uart1_rst().set_bit(),
+                _ => panic!("Unknown Uart Device!"),
+            });
+        system
+            .perip_rst_en0()
+            .modify(|_, w| match T::uart_number() {
+                0 => w.uart_rst().clear_bit(),
+                1 => w.uart1_rst().clear_bit(),
+                _ => panic!("Unknown Uart Device!"),
+            });
```

Comment:
Ideally there should be `System::reset` but we don't have it and I don't ask you to introduce it ... just a reminder - not an action item